src/Juvix/Compiler/Core/Transformation/LambdaLifting.hs

```diff
@@ -51,6 +55,83 @@ lambdaLiftNode aboveBl top =
+              defs = toList (letr ^. letRecValues)
+              ndefs :: Int
+              ndefs = length defs
+              letRecBinders :: [Info]
+              letRecBinders = Info.getInfoBinders ndefs (letr ^. letRecInfo)
+              bl' :: BinderList Info
+              bl' = BL.prepend letRecBinders bl
+          topSyms :: [Symbol] <- forM defs (const freshSymbol)
+          liftedDefs <- mapM (lambdaLiftNode bl') defs
+          body' <- lambdaLiftNode bl' (letr ^. letRecBody)
+          let -- free vars in each let
+              recItemsFreeVars :: [[(Index, Info)]]
+              recItemsFreeVars = mapMaybe helper . toList . freeVarsSet <$> liftedDefs
+                where
+                  -- throw away variables bound in the letrec and shift others
+                  helper :: Var -> Maybe (Index, Info)
+                  helper v
+                    | idx < ndefs = Nothing
+                    | otherwise =
+                        let idx' = idx - ndefs
+                         in Just (idx', BL.lookup idx' bl)
+                    where
+                      idx = v ^. varIndex
+              -- replace calls to letrec items to a calls to the fresh top symbols
+              subsCalls :: Node -> Node
+              subsCalls =
+                substs
+                  ( reverse
+                      [ mkApps' (mkIdent' sym) (map (mkVar' . fst) fv)
+                        | (sym, fv) <- zipExact topSyms recItemsFreeVars
+                      ]
+                  )
+              declareTopSyms :: Sem r ()
+              declareTopSyms =
+                sequence_
+                  [ do
+                      let topBody = captureFreeVars fv (subsCalls b)
+                          argsInfo :: [ArgumentInfo]
+                          argsInfo = map (argumentInfoFromInfo . snd) fv
+                      registerIdentNode sym topBody
+                      registerIdent
+                        IdentifierInfo
+                          { _identifierSymbol = sym,
+                            _identifierName = Nothing,
+                            _identifierType = typeFromArgs argsInfo,
+                            _identifierArgsNum = length fv,
+                            _identifierArgsInfo = argsInfo,
+                            _identifierIsExported = False
+                          }
+                    | (sym, b, fv) <- zip3Exact topSyms liftedDefs recItemsFreeVars
+                  ]
+              letItems :: [Node]
+              letItems =
+                [ mkApps' (mkIdent' s) (map (mkVar' . fst) fv)
+                  | (s, fv) <- zipExact topSyms recItemsFreeVars
+                ]
+          declareTopSyms
+          let -- free variables in the lets and the body need to be shifted
+              -- because we are introducing binders.
+              -- the topmost let is shifted 0
+              -- the lowermost is shifted (len - 1)
+              -- the final body is not shifted
```

Comment:
I think there is something wrong with this shifting. I changed the letrec test and it fails. You maybe forgot that the values had extra n variables that were then removed? But `substs` should take care of that, so I don't know.